### PR TITLE
Fix (core/SpellEffects): Don't randomize summoned guardian position if target is in the DB

### PR DESCRIPTION
**Pull request by kjack9**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

Adds an extra condition to the position randomization function for summoned guardian. This condition ensures that if the target for the summon is from the DB (`spell_target_position`) that it will appear exactly at that position every time.

## Issues Addressed:

- Needed for #17398
- Related to #12257
- Related to https://github.com/chromiecraft/chromiecraft/issues/3709

## SOURCE:

The changes have been validated through:
- [X] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

Before applying the PR:
1. `.go xyz 4032.49 -3357.65 115.06 329 1.52`
2. Engage Baron Rivendare (10440)
3. Observe that the first wave of skeletons appear exactly at the spell effect on each bone pile
4. Observe that subsequent waves appear in a radius around this point, sometimes causing skeletons to be summoned outside the room (through the wall)
5. Observe that the corpses from the skeletons are in different positions each time (because they are being summoned into random positions around the target position)

image

After applying the PR:
1. `.go xyz 4032.49 -3357.65 115.06 329 1.52`
2. Engage Baron Rivendare (10440)
3. Observe that all waves of skeletons appear exactly at the spell effect on each bone pile, in the same spot each time. None of them disappear into the wall.
4. Observe that the corpses from the skeletons overlap near-perfectly because they are being summoned to the exact target position each time

image

## Known Issues and TODO List:

None

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Spells/SpellEffects.cpp [hunk 1]
```diff
@@ -6058,7 +6058,12 @@ void Spell::SummonGuardian(uint32 i, uint32 entry, SummonPropertiesEntry const*
 
         // xinef: do not use precalculated position for effect summon pet in this function
         // it means it was cast by NPC and should have its position overridden
-        if (totalNumGuardians == 1 && GetSpellInfo()->Effects[i].Effect != SPELL_EFFECT_SUMMON_PET)
+        // unless the target is a position specified in the DB
+        if
+        (
+            (totalNumGuardians == 1 && GetSpellInfo()->Effects[i].Effect != SPELL_EFFECT_SUMMON_PET) ||
+            (!GetSpellInfo()->Effects[i].HasRadius() && GetSpellInfo()->Effects[i].TargetA.GetTarget() == TARGET_DEST_DB)
+        )
```
Comment (Nefertumm) [inline src/server/game/Spells/SpellEffects.cpp:6066]:
```suggestion
        if ((totalNumGuardians == 1 && GetSpellInfo()->Effects[i].Effect != SPELL_EFFECT_SUMMON_PET) ||
            (!GetSpellInfo()->Effects[i].HasRadius() && GetSpellInfo()->Effects[i].TargetA.GetTarget() == TARGET_DEST_DB))
```

Comment (kjack9) [reply src/server/game/Spells/SpellEffects.cpp]: I updated the format but added a new condition for a zero-radius spells. Trying to walk the line between compact and unreadable. I still prefer the original format (I think breaking it out is way easier to read), but let me know what you think of this hybrid approach.

File: src/server/game/Spells/SpellEffects.cpp [hunk 2]
```diff
@@ -6057,8 +6057,12 @@ void Spell::SummonGuardian(uint32 i, uint32 entry, SummonPropertiesEntry const*
...
 
         // xinef: do not use precalculated position for effect summon pet in this function
-        // it means it was cast by NPC and should have its position overridden
-        if (totalNumGuardians == 1 && GetSpellInfo()->Effects[i].Effect != SPELL_EFFECT_SUMMON_PET)
+        // it means it was cast by NPC and should have its position overridden unless the
+        // target position is specified in the DB AND the effect has no or zero radius
+        if ((totalNumGuardians == 1 && GetSpellInfo()->Effects[i].Effect != SPELL_EFFECT_SUMMON_PET) ||
+            (GetSpellInfo()->Effects[i].TargetA.GetTarget() == TARGET_DEST_DB &&
+                (!GetSpellInfo()->Effects[i].HasRadius() || GetSpellInfo()->Effects[i].RadiusEntry->RadiusMax == 0))
+           )
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellEffects.cpp:6065]:
```suggestion
            (!GetSpellInfo()->Effects[i].HasRadius() || GetSpellInfo()->Effects[i].RadiusEntry->RadiusMax == 0)))
```

**Comment by kjack9** [comment]
Any questions or concerns remaining on this PR before it's merged? Should fix Baron and any other fight where the summoning spells are targeted in the DB. Let me know if there are any other changes you'd like to see.

**Review by elthehablo — APPROVED** [review]

**Comment by elthehablo** [comment]
approved. but am thinking whether this is fully backwards compatible. I cannot really imagine it would break something though

**Comment by kjack9** [comment]
> approved. but am thinking whether this is fully backwards compatible. I cannot really imagine it would break something though

I did think about that some. I think the worst case scenario is that some event was relying on summoned guardians being spread out when they're created if they're using the same spawn location over and over again. The only effect from that would be visual - they'd show up on top of each other. And if they REALLY want that behavior, the right move is to change the radius of the spell, not to rely on buggy behavior.

If it does show up in the issue list, punt it my way and I'll own getting it fixed.

Thanks for reviewing.

**Comment by elthehablo** [comment]
> > approved. but am thinking whether this is fully backwards compatible. I cannot really imagine it would break something though
> 
> I did think about that some. I think the worst case scenario is that some event was relying on summoned guardians being spread out when they're created if they're using the same spawn location over and over again. The only effect from that would be visual - they'd show up on top of each other. And if they REALLY want that behavior, the right move is to change the radius of the spell, not to rely on buggy behavior.
> 
> If it does show up in the issue list, punt it my way and I'll own getting it fixed.
> 
> Thanks for reviewing.

Well put :)

**Pull request merged**